pkg/csi/service/common/commonco/k8sorchestrator/topology.go

```diff
@@ -0,0 +1,475 @@
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to fetch node object with name %q. Error: %v", nodeInfo.NodeName, err)
+	}
+	// Create spec for CSINodeTopology.
+	csiNodeTopologySpec := &csinodetopologyv1alpha1.CSINodeTopology{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nodeInfo.NodeName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "v1",
+					Kind:       "Node",
+					Name:       nodeObj.Name,
+					UID:        nodeObj.UID,
+				},
+			},
+		},
+		Spec: csinodetopologyv1alpha1.CSINodeTopologySpec{
+			NodeID: nodeInfo.NodeID,
+		},
+	}
+	// Create CSINodeTopology CR for the node.
+	err = volTopology.csiNodeTopologyK8sClient.Create(ctx, csiNodeTopologySpec)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to create CSINodeTopology CR. Error: %+v", err)
+		}
+	} else {
+		log.Infof("Successfully created a CSINodeTopology instance for NodeName: %q", nodeInfo.NodeName)
+	}
+
+	timeoutSeconds := int64((time.Duration(getCSINodeTopologyWatchTimeoutInMin(ctx)) * time.Minute).Seconds())
+	watchCSINodeTopology, err := volTopology.csiNodeTopologyWatcher.Watch(metav1.ListOptions{
+		FieldSelector:  fields.OneTermEqualSelector("metadata.name", nodeInfo.NodeName).String(),
+		TimeoutSeconds: &timeoutSeconds,
+		Watch:          true,
+	})
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to watch on CSINodeTopology instance with name %q. Error: %+v", nodeInfo.NodeName, err)
+	}
+	defer watchCSINodeTopology.Stop()
+
+	// Check if status gets updated in the instance within the given timeout seconds.
+	for event := range watchCSINodeTopology.ResultChan() {
+		csiNodeTopologyInstance, ok := event.Object.(*csinodetopologyv1alpha1.CSINodeTopology)
+		if !ok {
+			log.Warnf("Received unidentified object - %+v", event.Object)
+			continue
+		}
+		if csiNodeTopologyInstance.Name != nodeInfo.NodeName {
+			continue
+		}
+		switch csiNodeTopologyInstance.Status.Status {
+		case csinodetopologyv1alpha1.CSINodeTopologySuccess:
+			accessibleTopology := make(map[string]string)
+			for _, label := range csiNodeTopologyInstance.Status.TopologyLabels {
+				accessibleTopology[label.Key] = label.Value
+			}
+			return accessibleTopology, nil
+		case csinodetopologyv1alpha1.CSINodeTopologyError:
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to retrieve topology information for Node: %q. Error: %q", nodeInfo.NodeName,
+				csiNodeTopologyInstance.Status.ErrorMessage)
+		}
+	}
+	return nil, logger.LogNewErrorCodef(log, codes.Internal,
+		"timed out while waiting for topology labels to be updated in %q CSINodeTopology instance.",
+		nodeInfo.NodeName)
+}
+
+// getCSINodeTopologyWatchTimeoutInMin returns the timeout for watching
+// on CSINodeTopology instances for any updates.
+// If environment variable NODEGETINFO_WATCH_TIMEOUT_MINUTES is set and
```

Comment:
`NODEGETINFO_WATCH_TIMEOUT_MINUTES` is not seen in any of deployment yamls. Can you confirm why ?

Reply:
I have a default value set for it in the code like a lot of our other env variables. If customer wants to increase the value as his/her node takes a longer time to come up, then they can include this env var in their deployment YAML with an increased value.

Reply:
Then, please include the `NODEGETINFO_WATCH_TIMEOUT_MINUTES` in deployment yaml with default values

Reply:
Done

Reply:
I used the latest YAML to deploy the daemonset after these changes and it worked.
Logs:
```
2021-07-28T19:23:12.848Z	INFO	logger/logger.go:41	Setting default log level to :"DEVELOPMENT"
2021-07-28T19:23:12.851Z	INFO	vsphere-csi/main.go:57	Version : v2.1.0-rc.1-959-gdf2e6d7c-dirty	{"TraceId": "930b8e91-d294-4d95-9000-b81f4f808ac5"}
2021-07-28T19:23:12.851Z	DEBUG	commonco/utils.go:102	Container orchestrator init params: {InternalFeatureStatesConfigInfo:{Name:internal-feature-states.csi.vsphere.vmware.com Namespace:vmware-system-csi} ServiceMode:node}	{"TraceId": "930b8e91-d294-4d95-9000-b81f4f808ac5"}
2021-07-28T19:23:12.888Z	INFO	logger/logger.go:41	Setting default log level to :"DEVELOPMENT"
2021-07-28T19:23:12.888Z	INFO	k8sorchestrator/k8sorchestrator.go:152	Initializing k8sOrchestratorInstance	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
2021-07-28T19:23:12.903Z	INFO	kubernetes/kubernetes.go:85	k8s client using in-cluster config	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
2021-07-28T19:23:12.907Z	INFO	kubernetes/kubernetes.go:341	Setting client QPS to 100.000000 and Burst to 100.	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
2021-07-28T19:23:13.696Z	INFO	k8sorchestrator/k8sorchestrator.go:258	New internal feature states values stored successfully: map[async-query-volume:false block-volume-snapshot:false csi-auth-check:true csi-migration:true improved-csi-idempotency:false improved-volume-topology:true online-volume-extend:true trigger-csi-fullsync:false]	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
2021-07-28T19:23:13.728Z	INFO	k8sorchestrator/k8sorchestrator.go:178	k8sOrchestratorInstance initialized	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
2021-07-28T19:23:13.728Z	INFO	service/driver.go:107	Configured: "csi.vsphere.vmware.com" with clusterFlavor: "VANILLA" and mode: "node"	{"TraceId": "ddb5dce6-9b9e-455f-b928-4ac1011157fd"}
time="2021-07-28T19:23:13Z" level=info msg="identity service registered"
time="2021-07-28T19:23:13Z" level=info msg="node service registered"
time="2021-07-28T19:23:13Z" level=info msg=serving endpoint="unix:///csi/csi.sock"
2021-07-28T19:23:13.740Z	INFO	k8sorchestrator/k8sorchestrator.go:484	configMapAdded: Internal feature state values from "internal-feature-states.csi.vsphere.vmware.com" stored successfully: map[async-query-volume:false block-volume-snapshot:false csi-auth-check:true csi-migration:true improved-csi-idempotency:false improved-volume-topology:true online-volume-extend:true trigger-csi-fullsync:false]	{"TraceId": "3ae6ec48-6e8d-4e71-a211-ba3a3fd1c363"}
2021-07-28T19:23:14.966Z	INFO	service/node.go:660	NodeGetInfo: called with args {XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:14.966Z	INFO	service/node.go:682	NodeGetInfo: MAX_VOLUMES_PER_NODE is set to 0	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:14.966Z	INFO	kubernetes/kubernetes.go:85	k8s client using in-cluster config	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:14.967Z	INFO	kubernetes/kubernetes.go:341	Setting client QPS to 100.000000 and Burst to 100.	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:14.967Z	INFO	kubernetes/kubernetes.go:85	k8s client using in-cluster config	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:14.967Z	INFO	kubernetes/kubernetes.go:341	Setting client QPS to 100.000000 and Burst to 100.	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:15.027Z	INFO	k8sorchestrator/topology.go:204	Topology service initiated successfully	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:15.062Z	INFO	k8sorchestrator/topology.go:312	Timeout is set to 1 minute(s)	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:15.068Z	INFO	k8sorchestrator/topology.go:271	Got another object - "k8s-master-61"	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
2021-07-28T19:23:15.069Z	INFO	service/node.go:753	NodeGetInfo response: node_id:"k8s-node-0378" accessible_topology:<segments:<key:"topology.kubernetes.io/region" value:"region-1" > segments:<key:"topology.kubernetes.io/zone" value:"zone-a" > > 	{"TraceId": "e1bb71ba-d1d2-4bea-8793-b70748677590"}
```